client_option.go

```diff
@@ -0,0 +1,63 @@
+package gax
+
+type ClientOption interface {
+	Resolve(*clientSettings)
+}
+
+type clientOptions []ClientOption
+
+func (opts clientOptions) Resolve(s *clientSettings) *clientSettings {
+	for _, opt := range opts {
+		opt.Resolve(s)
+	}
+	return s
+}
+
+type clientSettings struct {
+	apiName    string
+	apiVersion string
+	endpoint   string
+	scopes     []string
+}
+
+type withAPIName string
```

Comment:
Each of the option here requires a type-func-func triplet. No real need to change it now, but consider https://play.golang.org/p/WdaQgmpe-K